src/main/scala/chisel3/util/SparseVec.scala

```diff
@@ -0,0 +1,228 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util
+
+import chisel3._
+import chisel3.experimental.SourceInfo
+import chisel3.util.{log2Up, BitPat, Mux1H, PriorityMux}
+import chisel3.util.experimental.decode.{decoder, TruthTable}
+import chisel3.util.experimental.InlineInstance
+import scala.collection.SeqMap
+import scala.collection.mutable.{ArrayBuffer, LinkedHashMap, LinkedHashSet}
+
+private[this] object Detail {
+
+  // A passthrough module used to realize the readonly method.
+  class ReadOnlyModule[A <: Data](gen: A) extends RawModule with InlineInstance {
+    val in = IO(Flipped(gen))
+    val out = IO(gen)
+    out :<>= in
+  }
+
+  // TODO: This method produces sub-par error messages. Replace usages of this
+  // with first-class Chisel support once it lands.
+  def readOnly[A <: Data](
+    gen: A
+  ): A = {
+    val readOnlyModule = Module(new ReadOnlyModule(chiselTypeOf(gen)))
+    readOnlyModule.in :<>= gen
+    readOnlyModule.out
+  }
+
+}
+
+/** A sparse vector.  This should be created using the [[SparseVec$]] factory. */
+class SparseVec[A <: Data] private (size: Int, gen: A, mapping: Seq[(Int, A)]) {
```

Comment:
I'm a little reluctant to give up the class name `SparseVec` to a non-Data (since it would be a breaking change to try to make it a Record after the fact). I'm not saying we need to do the Data API right this moment, but maybe we adopt a slightly different API which is just factory methods that take the `mapping`, `addr`, and `Type` and just go ahead and do the dynamic lookup there?

Reply:
Moved this to extend `Record`.